File: sdk/go/internal/tx/pipeline.go
package tx

import (
	"context"
	"fmt"
	"time"
)

// PipelineConfig holds configuration for the TX pipeline.
type PipelineConfig struct {
	ChainID         string
	GasPrice        uint64
	ConfirmTimeout  time.Duration
	PollInterval    time.Duration
}

// DefaultPipelineConfig returns a PipelineConfig with sensible defaults.
func DefaultPipelineConfig(chainID string) PipelineConfig {
	return PipelineConfig{
		ChainID:        chainID,
		GasPrice:       DefaultGasPrice,
		ConfirmTimeout: DefaultConfirmTimeout,
		PollInterval:   DefaultPollInterval,
	}
}

// ExecuteTx executes the full 4-phase TX pipeline:
//   Phase 1 - Assembly: query account, build TxBody + AuthInfo + SignDoc
//   Phase 2 - Signing: sign the SignDoc
//   Phase 3 - Broadcast: encode TxRaw and broadcast
//   Phase 4 - Confirmation: poll for TX inclusion
func ExecuteTx(ctx context.Context, querier ChainQuerier, signer Signer, cfg PipelineConfig, req TxRequest) (*TxResult, error) {
	// Phase 1: Assembly
	address := signer.GetAddress()
	pubKey, err := signer.GetPubKey()
	if err != nil {
		return nil, fmt.Errorf("getting public key: %w", err)
	}

	accountNumber, sequence, err := querier.GetAccountInfo(ctx, address)
	if err != nil {
		return nil, fmt.Errorf("getting account info for %s: %w", address, err)
	}

	// Determine gas limit
	gasLimit := req.GasLimit
	if gasLimit == 0 {
		gasLimit = DefaultGasForMessage(req.Message.TypeURL())
	}

	// Determine fee
	feeAmount := req.FeeAmount
	if feeAmount == 0 {
		gasPrice := cfg.GasPrice
		if gasPrice == 0 {
			gasPrice = DefaultGasPrice
		}
		feeAmount = CalculateFee(gasLimit, gasPrice)
	}

	feeDenom := req.FeeDenom
	if feeDenom == "" {
		feeDenom = DefaultFeeDenom
	}

	// Encode TxBody
	bodyBytes := EncodeTxBody([]MessageEncoder{req.Message}, req.Memo, req.TimeoutHeight)

	// Encode AuthInfo
	feeCoins := []Coin{{Denom: feeDenom, Amount: fmt.Sprintf("%d", feeAmount)}}
	authInfoBytes := EncodeAuthInfo(pubKey, sequence, feeCoins, gasLimit)

	// Encode SignDoc
	signDocBytes := EncodeSignDoc(bodyBytes, authInfoBytes, cfg.ChainID, accountNumber)

	// Phase 2: Signing
	signature, err := signer.Sign(signDocBytes)
	if err != nil {
		return nil, fmt.Errorf("signing transaction: %w", err)
	}

	// Phase 3: Broadcast
	txRawBytes := EncodeTxRaw(bodyBytes, authInfoBytes, signature)

	txHash, code, rawLog, err := querier.BroadcastTxSync(ctx, txRawBytes)
	if err != nil {
		return nil, fmt.Errorf("broadcasting transaction: %w", err)
	}

	// Check broadcast result code
	if code != 0 {
		return &TxResult{
			TxHash: txHash,
			Code:   code,
			RawLog: rawLog,
		}, fmt.Errorf("broadcast failed with code %d: %s", code, rawLog)
	}

	// Phase 4: Confirmation
	result, err := PollTxConfirmation(ctx, querier, txHash, cfg.ConfirmTimeout, cfg.PollInterval)
	if err != nil {
		// Return partial result with TX hash even on timeout
		return &TxResult{
			TxHash: txHash,
		}, fmt.Errorf("confirming transaction: %w", err)
	}

	return result, nil
}

File: sdk/csharp/src/Internal/Tx/Pipeline.cs
using System.Text.Json;
using Seocheon.Sdk.Errors;
using Seocheon.Sdk.Infrastructure;
using Seocheon.Sdk.Infrastructure.Signing;

namespace Seocheon.Sdk.Internal.Tx;

/// <summary>
/// 4-phase transaction pipeline: Build → Sign → Broadcast → Confirm.
/// </summary>
public static class Pipeline
{
    /// <summary>
    /// Executes a full transaction lifecycle.
    /// </summary>
    public static async Task<TxResult> ExecuteTx(
        IChainClient client,
        ISigningService signer,
        PipelineConfig config,
        TxRequest request,
        CancellationToken ct = default)
    {
        // Phase 1: Build
        var accountInfo = await client.GetAccountInfo(signer.GetAddress(), ct);
        var gasLimit = Gas.ResolveGasLimit(request.MessageTypeUrl, request.GasLimit);
        var feeAmount = request.FeeAmount > 0
            ? request.FeeAmount
            : Gas.CalculateFee(gasLimit, config.GasPrice);
        var feeDenom = string.IsNullOrEmpty(request.FeeDenom) ? "uppyeo" : request.FeeDenom;

        var messageAny = Envelope.EncodeAny(request.MessageTypeUrl, request.Message);
        var bodyBytes = Envelope.EncodeTxBody(messageAny, request.Memo, request.TimeoutHeight);
        var authInfoBytes = Envelope.EncodeAuthInfo(
            signer.GetPubKey(),
            accountInfo.Sequence,
            gasLimit,
            feeAmount,
            feeDenom
        );

        // Phase 2: Sign
        byte[] signature;
        try
        {
            var signDoc = Envelope.EncodeSignDoc(
                bodyBytes,
                authInfoBytes,
                config.ChainId,
                accountInfo.AccountNumber
            );
            signature = signer.Sign(signDoc);
        }
        catch (SdkException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw SdkErrors.SigningFailed(ex.Message, ex);
        }

        // Phase 3: Broadcast
        var txRaw = Envelope.EncodeTxRaw(bodyBytes, authInfoBytes, signature);
        var broadcastResult = await client.BroadcastTx(txRaw, "sync", ct);

        var txHash = broadcastResult.TxHash;
        if (broadcastResult.Code != 0)
            throw SdkErrors.BroadcastFailed($"code={broadcastResult.Code}: {broadcastResult.RawLog}");

        // Phase 4: Confirm
        return await ConfirmTx(client, txHash, config.ConfirmTimeout, config.PollInterval, ct);
    }

    private static async Task<TxResult> ConfirmTx(
        IChainClient client,
        string txHash,
        TimeSpan timeout,
        TimeSpan pollInterval,
        CancellationToken ct)
    {
        using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        cts.CancelAfter(timeout);

        while (!cts.IsCancellationRequested)
        {
            try
            {
                var txResponse = await client.GetTx(txHash, cts.Token);
                if (txResponse != null)
                {
                    if (txResponse.Code != 0)
                        throw SdkErrors.AbciCodeToError(txResponse.Code, txResponse.RawLog);

                    return new TxResult
                    {
                        TxHash = txHash,
                        Height = txResponse.Height,
                        Code = txResponse.Code,
                        GasUsed = txResponse.GasUsed,
                        GasWanted = txResponse.GasWanted,
                        RawLog = txResponse.RawLog,
                        Events = txResponse.Events
                    };
                }
            }
            catch (SdkException)
            {
                throw;
            }
            catch
            {
                // TX not yet indexed, continue polling
            }

            try
            {
                await Task.Delay(pollInterval, cts.Token);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }

        throw SdkErrors.TxTimeout(txHash);
    }
}

/// <summary>
/// Broadcast response from the chain.
/// </summary>
public sealed record BroadcastResponse
{
    public string TxHash { get; init; } = "";
    public uint Code { get; init; }
    public string RawLog { get; init; } = "";
}

/// <summary>
/// Transaction response from GetTx query.
/// </summary>
public sealed record TxResponse
{
    public string TxHash { get; init; } = "";
    public long Height { get; init; }
    public uint Code { get; init; }
    public ulong GasUsed { get; init; }
    public ulong GasWanted { get; init; }
    public string RawLog { get; init; } = "";
    public IReadOnlyList<TxEventData> Events { get; init; } = [];
}
